### PR TITLE
Fixed #25068 -- Adding migrations support for models with metaclasses.

### PR DESCRIPTION
**Pull request by MikeAmy**
Original ticket: https://code.djangoproject.com/ticket/25068
First proposed fix: https://github.com/django/django/pull/4968

I also encountered the same problem as kosz85. 
I have included the relevant test from the pull request 4968. 
It seems to make no difference to the test-run times compared to master.

Happy new year!

### CODE REVIEW

File: django/db/migrations/state.py [hunk 1]
```diff
@@ -575,25 +575,51 @@ def render(self, apps):
...
+        for base in self.bases:
+            if isinstance(base, six.string_types):
+                try:
+                    base = apps.get_model(base)
+                except LookupError:
+                    raise InvalidBasesError(
+                        "Cannot resolve one or more bases from %r" % (
+                            self.bases,
+                        )
+                    )
```
Comment (charettes) [inline django/db/migrations/state.py]: Now that we know exactly which base lookup failed we could adjust the `InvalidBasesError` message.

File: tests/migrations/test_state.py [hunk 2]
```diff
@@ -908,6 +909,37 @@ class Meta:
...
+
+        class ItermediateMetaClass(ModelBase, DummyMetaClass):
+            pass
+
+        # Create the same Model without metaclassmaker as normal.
+        class MetaclassTestModel(six.with_metaclass(ItermediateMetaClass, models.Model, Mixin)):
+            class Meta:
+                app_label = "migrations"
+                apps = new_apps
+                swappable = 'TEST_SWAPPABLE_MODEL'
```
Comment (charettes) [inline tests/migrations/test_state.py]: Is there a reason `swappable` is specified here? If it's the case a comment wouldn't hurt.

Comment (MikeAmy) [inline tests/migrations/test_state.py]: Well spotted. I had simply reused the test provided in kosz85's original PR. 

So, I've run the tests without swappable or my fix, and they failed, then I ran them without swappable but with my fix, and they passed. This leads me to conclude that swappable is not needed by the test. I've removed it.

File: django/db/migrations/state.py (hunk 1 above)
Comment (MikeAmy) [inline django/db/migrations/state.py]: Good idea, I've changed that error message.

File: tests/migrations/test_state.py [hunk 3]
```diff
@@ -908,6 +909,36 @@ class Meta:
...
+
+        class IntermediateMetaClass(ModelBase, DummyMetaClass):
+            pass
+
+        # Create the same Model without metaclassmaker as normal.
+        class MetaclassTestModel(six.with_metaclass(IntermediateMetaClass, models.Model, Mixin)):
+            class Meta:
+                app_label = "migrations"
+                apps = new_apps
+
```
Comment (MikeAmy) [inline tests/migrations/test_state.py]: Removed swappable. By running the tests with and without it, and with and without the fix, I determined it was not necessary.

File: django/db/migrations/state.py [hunk 4]
```diff
@@ -575,25 +575,49 @@ def render(self, apps):
...
+        bases = list()
+        meta_class_bases = list()
+        meta_class_seen_supertypes = set()
+        for base in self.bases:
+            if isinstance(base, six.string_types):
+                try:
+                    base = apps.get_model(base)
+                except LookupError:
+                    raise InvalidBasesError(
+                        "Cannot resolve %s base %s" % (self.name, base)
```
Comment (MikeAmy) [inline django/db/migrations/state.py]: Note that error message has been made more specific.

**Comment by MikeAmy** [comment]
I've found that the migrations take a long time to run for our 170 or so models, almost all of which have metaclasses, to inherit some common security/logging/concurrent update checking functionality. The tests take around 35 seconds to run the migrations before starting. After some simple profiling and reading the code, I believe the problem is that the algorithmic complexity of StateApps.render_multiple is O(n<sup>2</sup>), i.e. that this is a more general problem. I guess this is a known issue.

**Comment by MikeAmy** [comment]
Referring to my comment above about complexity, I tried removing almost all of our metaclasses and it made no different to the time to run migrations.

File: django/db/migrations/state.py [hunk 5]
```diff
@@ -575,25 +575,49 @@ def render(self, apps):
...
         # Then, work out our bases
-        try:
-            bases = tuple(
-                (apps.get_model(base) if isinstance(base, six.string_types) else base)
-                for base in self.bases
-            )
-        except LookupError:
-            raise InvalidBasesError("Cannot resolve one or more bases from %r" % (self.bases,))
+        bases = list()
+        meta_class_bases = list()
```
Comment (loic) [inline django/db/migrations/state.py]: Nitpick but metaclass should be one word.

**Comment by timgraham** [comment]
Closing as per "[not the right approach](https://code.djangoproject.com/ticket/25068#comment:25)" comment on the ticket.

**Pull request closed without merging**